File: server/services/database-config.ts
import fs from "fs";
import path from "path";
import { Pool } from "pg";

interface DatabaseConnection {
  host: string;
  port: number;
  database: string;
  username: string;
  password: string;
  type: string;
  isActive: boolean;
  schema: any;
  lastConnected: string | null;
}

interface DatabaseConfig {
  users: Record<string, {
    currentConnection: string | null;
    connections: Record<string, DatabaseConnection>;
  }>;
}

const CONFIG_PATH = path.join(process.cwd(), "server/config/database.json");

export class DatabaseConfigService {
  private static instance: DatabaseConfigService;
  private config!: DatabaseConfig;

  constructor() {
    this.loadConfig();
  }

  static getInstance(): DatabaseConfigService {
    if (!DatabaseConfigService.instance) {
      DatabaseConfigService.instance = new DatabaseConfigService();
    }
    return DatabaseConfigService.instance;
  }

  private loadConfig(): void {
    try {
      const configData = fs.readFileSync(CONFIG_PATH, "utf8");
      this.config = JSON.parse(configData);
      
      // Migrate old format to new user-specific format if needed
      if ('currentConnection' in this.config && 'connections' in this.config) {
        const oldConfig: any = this.config;
        this.config = {
          users: {
            'migration': {
              currentConnection: oldConfig.currentConnection,
              connections: oldConfig.connections
            }
          }
        };
        this.saveConfig();
      }
    } catch (error) {
      // If config doesn't exist, create default with user-specific structure
      this.config = {
        users: {}
      };
      this.saveConfig();
    }
  }

  private saveConfig(): void {
    try {
      fs.writeFileSync(CONFIG_PATH, JSON.stringify(this.config, null, 2));
    } catch (error) {
      console.error("Failed to save database config:", error);
      throw new Error("Could not save database configuration");
    }
  }

  async testConnection(connectionParams: {
    host: string;
    port: number;
    database: string;
    username: string;
    password: string;
  }): Promise<boolean> {
    console.log("Testing database connection with params:", {
      host: connectionParams.host,
      port: connectionParams.port,
      database: connectionParams.database,
      username: connectionParams.username,
      password: '***'
    });
    
    // Check if this is a cloud database that requires SSL
    const requiresSSL = connectionParams.host.includes('neon.tech') || 
                       connectionParams.host.includes('supabase.') ||
                       connectionParams.host.includes('amazonaws.com') ||
                       connectionParams.host.includes('planetscale.') ||
                       connectionParams.host.includes('railway.');

    const pool = new Pool({
      host: connectionParams.host,
      port: connectionParams.port,
      database: connectionParams.database,
      user: connectionParams.username,
      password: connectionParams.password,
      ssl: requiresSSL ? { rejectUnauthorized: false } : false,
      connectionTimeoutMillis: 30000,
      idleTimeoutMillis: 30000,
      max: 1, // Only one connection for testing
      allowExitOnIdle: true,
    });

    let client;
    try {
      client = await pool.connect();
      await client.query("SELECT 1");
      return true;
    } catch (error) {
      console.error("Database connection test failed:", error);
      return false;
    } finally {
      if (client) {
        client.release();
      }
      try {
        await pool.end();
      } catch (endError) {
        console.error("Error ending pool during connection test:", endError);
      }
    }
  }

  async extractSchema(connectionParams: {
    host: string;
    port: number;
    database: string;
    username: string;
    password: string;
  }): Promise<any> {
    // Check if this is a cloud database that requires SSL
    const requiresSSL = connectionParams.host.includes('neon.tech') || 
                       connectionParams.host.includes('supabase.') ||
                       connectionParams.host.includes('amazonaws.com') ||
                       connectionParams.host.includes('planetscale.') ||
                       connectionParams.host.includes('railway.');

    const pool = new Pool({
      host: connectionParams.host,
      port: connectionParams.port,
      database: connectionParams.database,
      user: connectionParams.username,
      password: connectionParams.password,
      ssl: requiresSSL ? { rejectUnauthorized: false } : false,
      connectionTimeoutMillis: 30000,
      idleTimeoutMillis: 30000,
      max: 1, // Only one connection for schema extraction
      allowExitOnIdle: true,
    });

    let client;
    try {
      client = await pool.connect();

      // Get all tables in the database
      const tablesQuery = `
        SELECT 
          table_name,
          table_schema
        FROM information_schema.tables 
        WHERE table_schema NOT IN ('information_schema', 'pg_catalog')
        ORDER BY table_name;
      `;

      const tablesResult = await client.query(tablesQuery);
      const tables = [];

      for (const tableRow of tablesResult.rows) {
        const tableName = tableRow.table_name;
        const tableSchema = tableRow.table_schema;

        // Get columns for each table
        const columnsQuery = `
          SELECT 
            column_name,
            data_type,
            is_nullable,
            column_default,
            character_maximum_length,
            numeric_precision,
            numeric_scale
          FROM information_schema.columns 
          WHERE table_name = $1 AND table_schema = $2
          ORDER BY ordinal_position;
        `;

        const columnsResult = await client.query(columnsQuery, [
          tableName,
          tableSchema,
        ]);

        tables.push({
          name: tableName,
          schema: tableSchema,
          columns: columnsResult.rows.map((col) => ({
            name: col.column_name,
            type: col.data_type,
            nullable: col.is_nullable === "YES",
            default: col.column_default,
            maxLength: col.character_maximum_length,
            precision: col.numeric_precision,
            scale: col.numeric_scale,
          })),
        });
      }

      // Format schema to match the existing format expected by the frontend
      const formattedTables: Record<string, any> = {};
      tables.forEach((table: any) => {
        const columns: Record<string, any> = {};
        table.columns.forEach((col: any) => {
          columns[col.name] = {
            type: col.type,
            nullable: col.nullable,
            default: col.default,
          };
        });
        formattedTables[table.name] = { columns };
      });

      return {
        tables: formattedTables,
        extractedAt: new Date().toISOString(),
        totalTables: tables.length,
        rawTables: tables, // Keep original format for reference
      };
    } catch (error) {
      console.error("Schema extraction failed:", error);
      throw new Error("Failed to extract database schema");
    } finally {
      if (client) {
        client.release();
      }
      try {
        await pool.end();
      } catch (endError) {
        console.error("Error ending pool during schema extraction:", endError);
      }
    }
  }

  private findExistingConnection(userId: string, connectionParams: {
    host: string;
    port: number;
    database: string;
    username: string;
  }): string | null {
    if (!this.config.users[userId]?.connections) return null;

    // Look for existing connection with same host, port, database, and username
    for (const [connectionId, connection] of Object.entries(this.config.users[userId].connections)) {
      if (
        connection.host === connectionParams.host &&
        connection.port === connectionParams.port &&
        connection.database === connectionParams.database &&
        connection.username === connectionParams.username
      ) {
        return connectionId;
      }
    }
    return null;
  }

  async addConnection(
    userId: string,
    connectionId: string,
    connectionParams: {
      host: string;
      port: number;
      database: string;
      username: string;
      password: string;
    },
  ): Promise<{ success: boolean; schema?: any; error?: string; isExisting?: boolean; existingConnectionId?: string }> {
    try {
      // Test the connection first
      const isValid = await this.testConnection(connectionParams);
      if (!isValid) {
        return {
          success: false,
          error: "Failed to connect to database with provided credentials",
        };
      }

      // Initialize user config if it doesn't exist
      if (!this.config.users[userId]) {
        this.config.users[userId] = {
          currentConnection: null,
          connections: {}
        };
      }

      // Check for existing connection with same database parameters
      const existingConnectionId = this.findExistingConnection(userId, connectionParams);
      
      if (existingConnectionId) {
        // Update existing connection instead of creating duplicate
        console.log(`Found existing connection for ${connectionParams.database} on ${connectionParams.host}, updating it instead of creating duplicate`);
        
        // Deactivate all connections for this user
        Object.keys(this.config.users[userId].connections).forEach((key) => {
          this.config.users[userId].connections[key].isActive = false;
        });

        // Update existing connection with fresh schema and activate it
        console.log(`Extracting fresh schema for existing connection ${existingConnectionId}`);
        const schema = await this.extractSchema(connectionParams);
        console.log(`Fresh schema extracted with ${Object.keys(schema.tables || {}).length} tables`);
        this.config.users[userId].connections[existingConnectionId] = {
          ...connectionParams,
          type: "postgresql",
          isActive: true,
          schema: schema,
          lastConnected: new Date().toISOString(),
        };

        this.config.users[userId].currentConnection = existingConnectionId;
        this.saveConfig();

        return { success: true, schema, isExisting: true, existingConnectionId };
      }

      // Extract schema for new connection
      console.log(`Extracting fresh schema for new connection ${connectionId}`);
      const schema = await this.extractSchema(connectionParams);
      console.log(`Fresh schema extracted with ${Object.keys(schema.tables || {}).length} tables`);

      // Deactivate current active connection for this user
      Object.keys(this.config.users[userId].connections).forEach((key) => {
        this.config.users[userId].connections[key].isActive = false;
      });

      // Add new connection for this user
      this.config.users[userId].connections[connectionId] = {
        ...connectionParams,
        type: "postgresql",
        isActive: true,
        schema: schema,
        lastConnected: new Date().toISOString(),
      };

      this.config.users[userId].currentConnection = connectionId;
      this.saveConfig();

      return { success: true, schema };
    } catch (error) {
      console.error("Failed to add database connection:", error);
      return {
        success: false,
        error:
          error instanceof Error ? error.message : "Unknown error occurred",
      };
    }
  }

  getCurrentConnection(userId: string): DatabaseConnection | null {
    if (!this.config.users[userId]?.currentConnection) return null;
    return this.config.users[userId].connections[this.config.users[userId].currentConnection!] || null;
  }

  getConnectionSchema(userId: string): any {
    const current = this.getCurrentConnection(userId);
    return current?.schema || null;
  }

  getAllConnections(userId: string): Record<string, DatabaseConnection> {
    if (!this.config.users[userId]?.connections) return {};
    
    // Clean up duplicates before returning connections
    this.cleanupDuplicateConnections(userId);
    return this.config.users[userId]?.connections || {};
  }

  private cleanupDuplicateConnections(userId: string): void {
    if (!this.config.users[userId]?.connections) return;

    const connections = this.config.users[userId].connections;
    const seen = new Set<string>();
    const toRemove: string[] = [];

    // Build unique identifier for each connection and track duplicates
    for (const [connectionId, connection] of Object.entries(connections)) {
      const identifier = `${connection.host}:${connection.port}:${connection.database}:${connection.username}`;
      
      if (seen.has(identifier)) {
        // This is a duplicate - mark for removal (keep the first occurrence)
        toRemove.push(connectionId);
        console.log(`Found duplicate connection: ${connectionId} for database ${connection.database}`);
      } else {
        seen.add(identifier);
      }
    }

    // Remove duplicates
    let hasChanges = false;
    for (const connectionId of toRemove) {
      delete this.config.users[userId].connections[connectionId];
      hasChanges = true;
      
      // If this was the current connection, find a replacement
      if (this.config.users[userId].currentConnection === connectionId) {
        const remainingConnections = Object.keys(this.config.users[userId].connections);
        if (remainingConnections.length > 0) {
          this.config.users[userId].currentConnection = remainingConnections[0];
          this.config.users[userId].connections[remainingConnections[0]].isActive = true;
        } else {
          this.config.users[userId].currentConnection = null;
        }
      }
    }

    // Save config if we made changes
    if (hasChanges) {
      this.saveConfig();
      console.log(`Cleaned up ${toRemove.length} duplicate connection(s) for user ${userId}`);
    }
  }

  setActiveConnection(userId: string, connectionId: string): boolean {
    if (!this.config.users[userId]?.connections[connectionId]) return false;

    // Deactivate all connections for this user
    Object.keys(this.config.users[userId].connections).forEach((key) => {
      this.config.users[userId].connections[key].isActive = false;
    });

    // Activate selected connection
    this.config.users[userId].connections[connectionId].isActive = true;
    this.config.users[userId].currentConnection = connectionId;
    this.saveConfig();
    return true;
  }

  removeConnection(userId: string, connectionId: string): boolean {
    if (!this.config.users[userId]?.connections[connectionId]) return false;

    delete this.config.users[userId].connections[connectionId];

    if (this.config.users[userId].currentConnection === connectionId) {
      // Find another connection to set as current, or set to null
      const remainingConnections = Object.keys(this.config.users[userId].connections);
      if (remainingConnections.length > 0) {
        const newCurrentConnection = remainingConnections[0];
        this.config.users[userId].currentConnection = newCurrentConnection;
        this.config.users[userId].connections[newCurrentConnection].isActive = true;
      } else {
        this.config.users[userId].currentConnection = null;
      }
    }

    this.saveConfig();
    return true;
  }
}

File: backend/services/database_config.py
import asyncio
import json
import os
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, Optional

import psycopg


# Resolve to backend/config/database.json regardless of CWD
CONFIG_PATH = Path(__file__).resolve().parents[1] / "config" / "database.json"


@dataclass
class DatabaseConnection:
    host: str
    port: int
    database: str
    username: str
    password: str
    type: str = "postgresql"
    isActive: bool = False
    schema: Optional[Dict[str, Any]] = None
    lastConnected: Optional[str] = None


class DatabaseConfigService:
    _instance: "DatabaseConfigService | None" = None

    def __init__(self) -> None:
        self.config: Dict[str, Any] = {}
        self._load_config()

    @classmethod
    def get_instance(cls) -> "DatabaseConfigService":
        if cls._instance is None:
            cls._instance = DatabaseConfigService()
        return cls._instance

    def _load_config(self) -> None:
        try:
            with open(CONFIG_PATH, "r", encoding="utf-8") as f:
                self.config = json.load(f)
            # migrate if needed
            if "currentConnection" in self.config and "connections" in self.config:
                self.config = {
                    "users": {
                        "migration": {
                            "currentConnection": self.config.get("currentConnection"),
                            "connections": self.config.get("connections", {}),
                        }
                    }
                }
                self._save_config()
        except Exception:
            self.config = {"users": {}}
            CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
            self._save_config()

    def _save_config(self) -> None:
        CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
        with open(CONFIG_PATH, "w", encoding="utf-8") as f:
            json.dump(self.config, f, indent=2)

    async def test_connection(self, params: Dict[str, Any]) -> bool:
        requires_ssl = any(
            host in params["host"]
            for host in ["neon.tech", "supabase.", "amazonaws.com", "planetscale.", "railway."]
        )
        conninfo = psycopg.conninfo.make_conninfo(
            host=params["host"],
            port=params["port"],
            dbname=params["database"],
            user=params["username"],
            password=params["password"],
            sslmode="require" if requires_ssl else "disable",
        )
        try:
            async with await psycopg.AsyncConnection.connect(conninfo) as aconn:
                async with aconn.cursor() as cur:
                    await cur.execute("SELECT 1")
                    await cur.fetchone()
            return True
        except Exception as e:
            print("Database connection test failed:", e)
            return False

    async def extract_schema(self, params: Dict[str, Any]) -> Dict[str, Any]:
        requires_ssl = any(
            host in params["host"]
            for host in ["neon.tech", "supabase.", "amazonaws.com", "planetscale.", "railway."]
        )
        conninfo = psycopg.conninfo.make_conninfo(
            host=params["host"],
            port=params["port"],
            dbname=params["database"],
            user=params["username"],
            password=params["password"],
            sslmode="require" if requires_ssl else "disable",
        )
        tables: list[dict[str, Any]] = []
        async with await psycopg.AsyncConnection.connect(conninfo) as aconn:
            async with aconn.cursor() as cur:
                await cur.execute(
                    """
                    SELECT table_name, table_schema
                    FROM information_schema.tables
                    WHERE table_schema NOT IN ('information_schema', 'pg_catalog')
                    ORDER BY table_name;
                    """
                )
                for table_name, table_schema in await cur.fetchall():
                    await cur.execute(
                        """
                        SELECT column_name, data_type, is_nullable, column_default,
                               character_maximum_length, numeric_precision, numeric_scale
                        FROM information_schema.columns
                        WHERE table_name = %s AND table_schema = %s
                        ORDER BY ordinal_position;
                        """,
                        (table_name, table_schema),
                    )
                    cols = await cur.fetchall()
                    columns = []
                    for (
                        column_name,
                        data_type,
                        is_nullable,
                        column_default,
                        character_maximum_length,
                        numeric_precision,
                        numeric_scale,
                    ) in cols:
                        columns.append(
                            {
                                "name": column_name,
                                "type": data_type,
                                "nullable": is_nullable == "YES",
                                "default": column_default,
                                "maxLength": character_maximum_length,
                                "precision": numeric_precision,
                                "scale": numeric_scale,
                            }
                        )
                    tables.append(
                        {"name": table_name, "schema": table_schema, "columns": columns}
                    )

        formatted: Dict[str, Any] = {}
        for table in tables:
            cols = {}
            for col in table["columns"]:
                cols[col["name"]] = {
                    "type": col["type"],
                    "nullable": col["nullable"],
                    "default": col["default"],
                }
            formatted[table["name"]] = {"columns": cols}

        return {
            "tables": formatted,
            "extractedAt": __import__("datetime").datetime.utcnow().isoformat(),
            "totalTables": len(tables),
            "rawTables": tables,
        }

    def _find_existing_connection(self, user_id: str, params: Dict[str, Any]) -> Optional[str]:
        if not self.config.get("users", {}).get(user_id, {}).get("connections"):
            return None
        for cid, conn in self.config["users"][user_id]["connections"].items():
            if (
                conn.get("host") == params["host"]
                and conn.get("port") == params["port"]
                and conn.get("database") == params["database"]
                and conn.get("username") == params["username"]
            ):
                return cid
        return None

    async def add_connection(
        self, user_id: str, connection_id: str, params: Dict[str, Any]
    ) -> Dict[str, Any]:
        is_valid = await self.test_connection(params)
        if not is_valid:
            return {
                "success": False,
                "error": "Failed to connect to database with provided credentials",
            }

        self.config.setdefault("users", {})
        self.config["users"].setdefault(user_id, {"currentConnection": None, "connections": {}})

        existing_id = self._find_existing_connection(user_id, params)
        if existing_id:
            for key in list(self.config["users"][user_id]["connections"].keys()):
                self.config["users"][user_id]["connections"][key]["isActive"] = False
            schema = await self.extract_schema(params)
            self.config["users"][user_id]["connections"][existing_id] = {
                **params,
                "type": "postgresql",
                "isActive": True,
                "schema": schema,
                "lastConnected": __import__("datetime").datetime.utcnow().isoformat(),
            }
            self.config["users"][user_id]["currentConnection"] = existing_id
            self._save_config()
            return {"success": True, "schema": schema, "isExisting": True, "existingConnectionId": existing_id}

        schema = await self.extract_schema(params)
        for key in list(self.config["users"][user_id]["connections"].keys()):
            self.config["users"][user_id]["connections"][key]["isActive"] = False
        self.config["users"][user_id]["connections"][connection_id] = {
            **params,
            "type": "postgresql",
            "isActive": True,
            "schema": schema,
            "lastConnected": __import__("datetime").datetime.utcnow().isoformat(),
        }
        self.config["users"][user_id]["currentConnection"] = connection_id
        self._save_config()
        return {"success": True, "schema": schema}

    def get_current_connection(self, user_id: str) -> Optional[Dict[str, Any]]:
        user = self.config.get("users", {}).get(user_id)
        if not user or not user.get("currentConnection"):
            return None
        return user["connections"].get(user["currentConnection"])

    def get_all_connections(self, user_id: str) -> Dict[str, Any]:
        self._cleanup_duplicates(user_id)
        return self.config.get("users", {}).get(user_id, {}).get("connections", {})

    def _cleanup_duplicates(self, user_id: str) -> None:
        user = self.config.get("users", {}).get(user_id)
        if not user:
            return
        connections = user.get("connections", {})
        seen = set()
        to_remove: list[str] = []
        for cid, conn in connections.items():
            identifier = f"{conn.get('host')}:{conn.get('port')}:{conn.get('database')}:{conn.get('username')}"
            if identifier in seen:
                to_remove.append(cid)
            else:
                seen.add(identifier)
        changed = False
        for cid in to_remove:
            connections.pop(cid, None)
            changed = True
            if user.get("currentConnection") == cid:
                remaining = list(connections.keys())
                user["currentConnection"] = remaining[0] if remaining else None
                if remaining:
                    connections[remaining[0]]["isActive"] = True
        if changed:
            self._save_config()

    def set_active_connection(self, user_id: str, connection_id: str) -> bool:
        user = self.config.get("users", {}).get(user_id)
        if not user or connection_id not in user.get("connections", {}):
            return False
        for key in list(user["connections"].keys()):
            user["connections"][key]["isActive"] = False
        user["connections"][connection_id]["isActive"] = True
        user["currentConnection"] = connection_id
        self._save_config()
        return True

    def remove_connection(self, user_id: str, connection_id: str) -> bool:
        user = self.config.get("users", {}).get(user_id)
        if not user or connection_id not in user.get("connections", {}):
            return False
        user["connections"].pop(connection_id, None)
        if user.get("currentConnection") == connection_id:
            remaining = list(user["connections"].keys())
            user["currentConnection"] = remaining[0] if remaining else None
            if remaining:
                user["connections"][remaining[0]]["isActive"] = True
        self._save_config()
        return True


